File: solutions/elixir/010-rotting-oranges.ex
# Problem 10: Rotting Oranges (LeetCode 994)
# Difficulty: Med
# Language: Elixir
# 
defmodule RottingOranges do
  def oranges_rotting(grid) do
    rows = length(grid)
    cols = length(hd(grid))
    cells = for {row, r} <- Enum.with_index(grid),
                {val, c} <- Enum.with_index(row), do: {{r, c}, val}
    rotten = for {{r, c}, 2} <- cells, do: {r, c}
    fresh = MapSet.new(for {{r, c}, 1} <- cells, do: {r, c})
    bfs(rotten, fresh, rows, cols, 0)
  end

  defp bfs([], fresh, _, _, time), do: if MapSet.size(fresh) == 0, do: time, else: -1
  defp bfs(queue, fresh, rows, cols, time) do
    {new_queue, new_fresh} =
      Enum.reduce(queue, {[], fresh}, fn {r, c}, {nq, fr} ->
        [{r-1,c},{r+1,c},{r,c-1},{r,c+1}]
        |> Enum.filter(fn {nr, nc} -> nr >= 0 and nr < rows and nc >= 0 and nc < cols end)
        |> Enum.reduce({nq, fr}, fn pos, {nq2, fr2} ->
          if MapSet.member?(fr2, pos),
            do: {[pos | nq2], MapSet.delete(fr2, pos)},
            else: {nq2, fr2}
        end)
      end)
    if new_queue == [],
      do: (if MapSet.size(new_fresh) == 0, do: time, else: -1),
      else: bfs(new_queue, new_fresh, rows, cols, time + 1)
  end
end

File: solutions/typescript/010-rotting-oranges.ts
/**
 * Problem 10: Rotting Oranges (LeetCode 994)
 * Difficulty: Med
 * Language: TypeScript
 */
function orangesRotting(grid: number[][]): number {
  const rows = grid.length, cols = grid[0].length;
  type State = { grid: number[][]; queue: [number, number][]; time: number };

  const neighbors = (r: number, c: number): [number, number][] =>
    [[r-1,c],[r+1,c],[r,c-1],[r,c+1]]
      .filter(([nr,nc]) => nr >= 0 && nr < rows && nc >= 0 && nc < cols) as [number,number][];

  const initial: State = {
    grid: grid.map(row => [...row]),
    queue: grid.flatMap((row, r) =>
      row.map((v, c) => v === 2 ? [r, c] as [number, number] : null).filter(Boolean) as [number, number][]),
    time: 0,
  };

  const bfs = (state: State): number => {
    if (state.queue.length === 0) {
      return state.grid.some(row => row.includes(1)) ? -1 : state.time;
    }
    const nextQueue: [number, number][] = [];
    const newGrid = state.grid.map(row => [...row]);
    state.queue.forEach(([r, c]) => {
      neighbors(r, c).forEach(([nr, nc]) => {
        if (newGrid[nr][nc] === 1) {
          newGrid[nr][nc] = 2;
          nextQueue.push([nr, nc]);
        }
      });
    });
    return nextQueue.length === 0
      ? (newGrid.some(row => row.includes(1)) ? -1 : state.time)
      : bfs({ grid: newGrid, queue: nextQueue, time: state.time + 1 });
  };

  return bfs(initial);
}
